### PR TITLE
Fix build breakage on linux

### PR DESCRIPTION
**Pull request by rsgowman**

### CODE REVIEW

File: Firestore/core/src/firebase/firestore/immutable/sorted_map.h [hunk 1]
```diff
@@ -143,6 +143,8 @@ class SortedMap : public impl::SortedMapBase {
         return SortedMap{array_.insert(key, value)};
       case Tag::Tree:
         return SortedMap{tree_.insert(key, value)};
+      default:
+        abort();
```
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/immutable/sorted_map.h]: OK. If the compiler won't behave I guess we have to do it :-).

Let's add a `FIRESTORE_UNREACHABLE()` to our assert header and make it call abort/assert or whatever. This way it's slightly more obvious why we're adding this.

File: Firestore/core/src/firebase/firestore/util/firebase_assert.h [hunk 2]
```diff
@@ -103,6 +103,11 @@
   FIREBASE_DEV_ASSERT_MESSAGE_WITH_EXPRESSION(expression, expression, \
                                               __VA_ARGS__)
 
+// Indicates an area of the code that cannot be reached (except possibly due to
+// undefined behaviour). The only reasonable thing to do in these cases is to
+// immediately abort.
+#define FIREBASE_UNREACHABLE() abort()
```
Comment (rsgowman) [inline Firestore/core/src/firebase/firestore/util/firebase_assert.h]: Alternatively, we could call FailAssert() for a friendlier error message... but since this should be used only for "impossible" conditions, I'm not too concerned about it.

**Comment by rsgowman** [comment]
eg.
sorted_map.h:172:3: error: control reaches end of non-void function [-Werror=return-type]

Also see the conversation up-thread.

tl;dr: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=53479

**Comment by var-const** [comment]
@rsgowman Sorry, I missed the "outdated" branch.

Hmm, it appears that GCC is actually correct, or at the very least, reasonable in this case. I presumed that the cast would result in undefined behavior, but it appears that it's not the case, actually.

File: Firestore/core/src/firebase/firestore/util/firebase_assert.h [hunk 3]
```diff
@@ -103,6 +103,11 @@
   FIREBASE_DEV_ASSERT_MESSAGE_WITH_EXPRESSION(expression, expression, \
                                               __VA_ARGS__)
 
+// Indicates an area of the code that cannot be reached (except possibly due to
+// undefined behaviour). The only reasonable thing to do in these cases is to
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/util/firebase_assert.h]: Nit: IIUC, there is no undefined behavior involved in this case. From the thread you linked:

> Enumerations are simply named constants, they don't affect the type, or codegen, or what the valid values of the type are (except where the values of the highest and lowest determine the range of valid values).

I'll slightly expand the example from the thread:
```cpp
enum class Foo { SomeValue = 1, }; // Underlying type is int
std::string Bar() {
  Foo foo = static_cast<Foo>(42); // This is well-defined
  switch (foo) {
  case SomeValue:
    return "foobar";
  }
  // Control reaches here, resulting in undefined behavior
}
```

Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/util/firebase_assert.h]: I've fixed up the comment slightly.
